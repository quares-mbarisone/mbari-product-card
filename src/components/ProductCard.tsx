
import { useProduct } from "../hooks/useProduct";
import React, { createContext, CSSProperties, JSX } from 'react';
import { Product, ProductContextProps, OnChangeArgs, InitialValues, ProductCartHandlers } from "../interfaces";

import styles from "../styles/styles.module.css";

export interface Props {
    product: Product;
    children: ( args: ProductCartHandlers ) => JSX.Element;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args: OnChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues
}

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext;

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }:Props) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (
        <Provider value={{
            counter,
            increaseBy,
            maxCount,
            product,
        }}>
            <div 
                className={ `${ styles.productCard } ${ className }`}
                style={ style }
            >  
                { 
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount,
                        product,
                        increaseBy,
                        reset,
                    }) 
                }
            </div>
        </Provider>
    )
}


