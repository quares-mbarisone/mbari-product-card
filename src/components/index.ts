import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductCartHOCProps } from '../interfaces';

import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';

export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';

export const ProductCard:ProductCartHOCProps = Object.assign( ProductCardHOC, {
    Buttons: ProductButtons,
    Image: ProductImage,
    Title: ProductTitle,
});

export default ProductCard;