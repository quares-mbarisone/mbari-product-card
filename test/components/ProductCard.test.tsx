import React from 'react';

import { ProductCard } from '../../src/components';
import { fireEvent, render, screen } from '@testing-library/react';
import { product1 } from '../data/products';

describe('Pruebas en <ProductCard/>', () => {

    test('debe de mostrar el componente correctamente', () => {
        render(
            <ProductCard product={ product1 }>
                {
                    () => (
                        <h1>Product Card</h1>
                    )
                }
            </ProductCard>
        );
        expect(screen.getByText('Product Card')).toMatchSnapshot();
    });

    test('debe de incrementar el contador', () => {
        render(
            <ProductCard product={ product1 }>
                {
                    ({count, increaseBy}) => (
                        <>
                            <h1>Product Card</h1>
                            <p>{ count }</p>
                            <button 
                                onClick={() => increaseBy(1)}
                                aria-label='increase-btn'
                            > + </button>
                        </>
                    )
                }
            </ProductCard>
        );
        
        expect(screen.getByRole('paragraph').innerHTML).toBe('0');

        const increaseBtn = screen.getByLabelText('increase-btn')
        fireEvent.click(increaseBtn)

        expect(screen.getByRole('paragraph').innerHTML).toBe('1');

    });

});
