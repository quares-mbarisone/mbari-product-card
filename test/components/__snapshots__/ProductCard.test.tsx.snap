// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ProductCard/> debe de mostrar el componente correctamente 1`] = `
<h1>
  Product Card
</h1>
`;
